### PR TITLE
feature(action): Support Vote Action

### PR DESCRIPTION
**Pull request by gnomadic**

### CODE REVIEW

File: src/actions/types.ts [hunk 1]
```diff
@@ -70,6 +70,30 @@ export type DelegateActionParams = {
   delegator?: Address
 }
 
+//https://github.com/OpenZeppelin/openzeppelin-contracts/blob/e5c63635e3508a8d9d0afed091578cc4bb59a9c7/contracts/governance/IGovernor.sol#L142C1-L157C7
```
Comment (Quazia) [inline src/actions/types.ts:73]: These need to be generalized - so I wouldn't make it quite so specific to this contract. Needs to be the params we'd use for a vote action on _any_ protocol.

File: src/actions/types.ts [hunk 2]
```diff
@@ -70,6 +70,13 @@ export type DelegateActionParams = {
   delegator?: Address
 }
 
+export type VoteActionParams = {
```
Comment (gnomadic) [inline src/actions/types.ts:73]: proposalId and support are parameters to the vote functions on the contracts.

Just to confirm, we pull these params from the functions, not the events those functions emit right?

File: src/actions/types.ts [hunk 3]
```diff
@@ -70,6 +70,13 @@ export type DelegateActionParams = {
...
+  chainId: number
+  governorAddress?: Address
+  proposalId?: bigint | FilterOperator
+  support?: bigint | FilterOperator
```
Comment (gnomadic) [inline src/actions/types.ts:77]: It feels undemocratic to let people reward others who vote a specific way... should we support this?

File: src/actions/types.ts (hunk 1 above)
Comment (gnomadic) [reply src/actions/types.ts]: makes sense!  I'll try and find another implementation or two

Comment (gnomadic) [reply src/actions/types.ts]: I think we're good.

[ENS uses tally](https://www.tally.xyz/gov/ens), and[ ENS token uses openzeppelin](https://github.com/ensdomains/governance-contracts/blob/30a1989d89ab6b859b23db5430d345ef84208616/contracts/ENSToken.sol#L8)

Comment (gnomadic) [reply src/actions/types.ts]: PHEW!  nevermind!

https://github.com/llamaxyz/llama/blob/main/src/LlamaCore.sol#L434

File: src/actions/types.ts [hunk 4]
```diff
@@ -1,81 +1,82 @@
...
 
 export type VoteActionParams = {
-  chainId: number
-  governorAddress?: Address
-  proposalId?: bigint | FilterOperator
-  support?: bigint | FilterOperator
-}
+  chainId: number;
+  project: Address | string;
+  contractAddress: Address;
```
Comment (mmackz) [inline src/actions/types.ts]: Looks good to me.

I think `contractAddress` might work better as an optional parameter.

File: src/actions/types.ts [hunk 5]
```diff
@@ -1,81 +1,82 @@
...
+  proposalId?: bigint | FilterOperator;
+  support?: boolean | FilterOperator;
```
Comment (mmackz) [inline src/actions/types.ts]: afaik, the FilterOperator type is there so we can use operator functions like GreaterThan, etc. Im not sure if it is very useful here.

Comment (Quazia) [reply src/actions/types.ts]: 👍 Agreed with Mattie. I would only use the `FilterOperator` in places where it makes sense to use < > <= >= etc

**Review by Quazia — APPROVED** [review]
LGTM :shipit:

File: src/actions/types.ts [hunk 6]
```diff
@@ -70,6 +70,13 @@ export type DelegateActionParams = {
   delegator?: Address
 }
 
+export type VoteActionParams = {
+  chainId: number
+  project: Address | string
```
Comment (mmackz) [inline src/actions/types.ts:75]: nice selection of parameters, this should work well.

**Pull request merged**